src/models/train_model.py

```diff
@@ -0,0 +1,219 @@
+"""Module to train the GAN model"""
+
+from typing import Any
+
+import torch
+
+from src.models.losses import discriminator_loss, generator_loss, kl_loss
+from src.models.modules.discriminator import Discriminator
+from src.models.modules.generator import Generator
+from src.models.modules.image_encoder import InceptionEncoder, VGGEncoder
+from src.models.modules.text_encoder import TextEncoder
+from src.models.utils import (
+    copy_gen_params,
+    define_optimizers,
+    load_params,
+    prepare_labels,
+    save_image_and_caption,
+    save_model,
+)
+
+# pylint: disable=too-many-locals
+
+
+def train(data_loader: Any, config_dict: dict[str, Any]) -> None:
+    """
+    Function to train the GAN model
+    :param data_loader: Data loader for the dataset
+    :param vocab_len: Length of the vocabulary
+    :param config_dict: Dictionary containing the configuration parameters
+    """
+    (
+        Ng,  # pylint: disable=invalid-name
+        D,  # pylint: disable=invalid-name
+        condition_dim,
+        noise_dim,
+        disc_lr,
+        gen_lr,
+        batch_size,
+        device,
+        epochs,
+        vocab_len,
+        ix2word,
+        output_dir,
+        snapshot,
+        const_dict,
+    ) = (
+        config_dict["Ng"],
+        config_dict["D"],
+        config_dict["condition_dim"],
+        config_dict["noise_dim"],
+        config_dict["disc_lr"],
+        config_dict["gen_lr"],
+        config_dict["batch_size"],
+        config_dict["device"],
+        config_dict["epochs"],
+        config_dict["vocab_len"],
+        config_dict["ix2word"],
+        config_dict["output_dir"],
+        config_dict["snapshot"],
+        config_dict["const_dict"],
+    )
+
+    smooth_val_gen = const_dict["smooth_val_gen"]
+    lambda4 = const_dict["lambda4"]
+    generator = Generator(Ng, D, condition_dim, noise_dim).to(device)
+    discriminator = Discriminator().to(device)
+    text_encoder = TextEncoder(vocab_len, D, D // 2).to(device)
```

Comment:
is text encoder getting optimized? we're not using a pretrained encoder so you should initialize optimizer for it as well and maybe make it a part of discriminator or generator weights

Reply:
ohhh you're right... should have done that. Thanks for catching.